File: python/295_FindMedianfromDataStream.py
from unittest import TestCase
# https://leetcode.com/problems/find-median-from-data-stream
from heapq import *


class FindMedianfromDataStream:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.heaps = [], []

    def addNum(self, num):
        """
        :type num: int
        :rtype: void
        """
        lque, rque = self.heaps
        if len(lque) == 0 or num <= -lque[0]:
            heappush(lque, -num)
            if len(lque) - len(rque) > 1:
                heappush(rque, -heappop(lque))
        else:
            heappush(rque, num)
            if len(rque) > len(lque):
                heappush(lque, -heappop(rque))

    def findMedian(self):
        """
        :rtype: float
        """
        l, r = self.heaps
        return (-l[0]+r[0])/2 if len(l) == len(r) else -l[0]


class Tests(TestCase):
    def test1(self):
        med = FindMedianfromDataStream()
        med.addNum(1)
        med.addNum(2)
        self.assertEqual(1.5, med.findMedian())
        med.addNum(4)
        self.assertEqual(2, med.findMedian())

File: kotlin/src/main/kotlin/295.FindMedianfromDataStream.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/find-median-from-data-stream/


class FindMedianfromDataStream{
  /** initialize your data structure here. */
  val lque = PriorityQueue<Int>(reverseOrder())
  val rque = PriorityQueue<Int>()

  fun addNum(num: Int) {
    if(lque.isEmpty() || num <= lque.peek()) {
      lque.add(num)
      if(lque.size - rque.size > 1)
        rque.add(lque.poll())
    }else{
      rque.add(num)
      if(rque.size > lque.size)
        lque.add(rque.poll())
    }
  }

  fun findMedian(): Double = if(lque.size == rque.size) (lque.peek()+rque.peek())/2.0 else lque.peek()*1.0


  // todo: for condensed case e.g. among 1-100, use count_hash & linkedlist

  companion object {
    const val PRECISION = 1e-9
  }

  @Test
  fun test1(){
    addNum(1)
    addNum(2)
    assertEquals(1.5, findMedian(), PRECISION)
    addNum(3)
    assertEquals(2.0, findMedian(), PRECISION)
  }
}